flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/common/CommonPhysicalSink.scala

```diff
@@ -99,11 +105,33 @@ class CommonPhysicalSink (
 
         val operator = new SinkOperator(env.clean(sinkFunction), rowtimeFieldIndex, enforcer)
 
+        val inputParallelism = inputTransformation.getParallelism
+        val taskParallelism = env.getParallelism
+        val parallelism = if (runtimeProvider.isInstanceOf[ParallelismProvider]) runtimeProvider.asInstanceOf[ParallelismProvider].getParallelism.orElse(inputParallelism).intValue()
+        else inputParallelism
+
+        if (implicitly[Ordering[Int]].lteq(parallelism, 0)) throw new RuntimeException(s"the configured sink parallelism: $parallelism should not be less than zero or equal to zero")
+        if (implicitly[Ordering[Int]].gt(parallelism, taskParallelism)) throw new RuntimeException(s"the configured sink parallelism: $parallelism is larger than the task max parallelism: $taskParallelism")
+
+        val primaryKeys = TableSchemaUtils.getPrimaryKeyIndices(catalogTable.getSchema)
+        val containedRowKinds = changelogMode.getContainedKinds.toSet
+        val theFinalInputTransformation = if(inputParallelism == parallelism) inputTransformation //if the parallelism is not changed, do nothing
+        else (containedRowKinds, primaryKeys.toList) match {
+        // fixme : if rowKinds only contains  delete, is there somethinng to do with? Currently do nothing.
```

Comment:
Actually, this part should be under well discussed. 
For every changelog mode, from my perspective, should be treated differently. Now I only figured out what should be done in "UPSERT MODE" or `INSERT_ONLY` mode.

What shall we do for other changelog?
WDYT?

Reply:
Yes, you can just `changelogMode  == ChangelogMode.INSERT_ONLY` (scala, java should use equals), and do keyBy.

Reply:
@JingsongLi  

I only do `keyBy` on ChangelogMode which contains `update_after`, while on other changelogMode I just keep the transformation with do nothing upon it. Is that Proper or not?